src/solver/conv_asm_implicit_gemm_wrw_gtc_dynamic_xdlops.cpp

```diff
@@ -0,0 +1,437 @@
+}
+
+static inline float CallImplicitGemmWrwDynamic(const miopen::Handle& handle,
+                                               const conv::ProblemDescription& conv_problem,
+                                               ConstData_t src,
+                                               ConstData_t dst,
+                                               Data_t wei,
+                                               const std::vector<KernelInvoke>& kernels)
+{
+    float elapsed = 0.0f;
+
+    auto kernel = kernels[0];
+    // clang-format off
+    int hi           = conv_problem.GetOutHeight();
+    int wi           = conv_problem.GetOutWidth();
+    int n            = conv_problem.GetOutChannels();
+    int k            = conv_problem.GetInChannels();
+    int c            = conv_problem.GetInBatchSize();
+    int ho           = conv_problem.GetWeightsHeight();
+    int wo           = conv_problem.GetWeightsWidth();
+    int dilation_h   = conv_problem.GetInHeight() > 1 ? conv_problem.GetKernelStrideH() : 1;
+    int dilation_w   = conv_problem.GetInWidth() > 1 ? conv_problem.GetKernelStrideW() : 1;
+    int stride_h     = conv_problem.GetWeightsHeight() > 1? conv_problem.GetDilationH() : 1;
+    int stride_w     = conv_problem.GetWeightsWidth() > 1? conv_problem.GetDilationW() : 1;
+    int pad_h        = conv_problem.GetPadH();
+    int pad_w        = conv_problem.GetPadW();
+    int y            = conv_problem.GetInHeight();
+    int x            = conv_problem.GetInWidth();
+    int gemmk_groups = 0;
+    int GemmKPerBlock;
+
+    if((k % 128 == 0) && ((n * ho * wo) % 128 == 0))
+        GemmKPerBlock = 16;
+    else
+        GemmKPerBlock = 4;
+
+    gemmk_groups = GetImplicitGemmWrwGTCDynamicXdlopsGemmkGroups(conv_problem, GemmKPerBlock);
+
+    MIOPEN_LOG_I2(kernel.GetName() << " with groups for reduction: " << (1 << gemmk_groups) << " GemmKPerBlock: " << GemmKPerBlock);
+
+    // clang-format on
+    std::vector<OpKernelArg> opArgs;
+    opArgs.emplace_back(src);
+    opArgs.emplace_back(dst);
+    opArgs.emplace_back(wei);
+    opArgs.emplace_back(hi);
+    opArgs.emplace_back(wi);
+    opArgs.emplace_back(n);
+    opArgs.emplace_back(k);
+    opArgs.emplace_back(c);
+    opArgs.emplace_back(ho);
+    opArgs.emplace_back(wo);
+    opArgs.emplace_back(stride_h);
+    opArgs.emplace_back(stride_w);
+    opArgs.emplace_back(dilation_h);
+    opArgs.emplace_back(dilation_w);
+    opArgs.emplace_back(pad_h);
+    opArgs.emplace_back(pad_w);
+    opArgs.emplace_back(y);
+    opArgs.emplace_back(x);
+    opArgs.emplace_back(gemmk_groups);
+    kernel(opArgs);
+
+    if(handle.IsProfilingEnabled())
+        elapsed += handle.GetKernelTime();
+
+    return elapsed;
+}
+
+// find wrw dynamic kernel by a simple algo
+// check wether this kernel can be applicable
+static inline bool FindImplicitGemmWrwGTCDynamicXdlopsKernel(const ConvolutionContext& ctx,
+                                                        std::string& kernel_name,
```

Comment:
It would be a more modern way to return a `boost::optional<some_struct/tupple_with_data>`, than having a bool return with ref args. Also, you won't have to create unused variables for the ref args when you don't need them.

Reply:
ok, thanks. I will refine the code.

Reply:
@DrizztDoUrden I refactored this function.